src/intrusive_double_linked_list.rs

```diff
@@ -48,278 +50,335 @@ impl<T> DerefMut for ListNode<T> {
 /// of type `T`.
 #[derive(Debug)]
 pub struct LinkedList<T> {
-    head: *mut ListNode<T>,
-    tail: *mut ListNode<T>,
+    head: Option<NonNull<ListNode<T>>>,
+    tail: Option<NonNull<ListNode<T>>>,
 }
 
 impl<T> LinkedList<T> {
     /// Creates an empty linked list
     pub fn new() -> Self {
         LinkedList::<T> {
-            head: null_mut(),
-            tail: null_mut(),
+            head: None,
+            tail: None,
         }
     }
```

Comment:
_very_ minor nit: now that you're using `Option`s, you could avoid some boilerplate by deriving `Default`. just a note!

Reply:
I actually like that kind of initialization. Makes it easy to see what is going on.